Clarify initialization vector sizes in subtlecrypto docs

benjreinhart opened the pull request:
### Description

Clearly list the recommended initialization vector sizes for AES-CBC (16 bytes) and AES-GCM (12 bytes). The existing docs use the language "Must be 16 bytes," implying this is true for both cbc/gcm, but that is incorrect.

### Motivation

To be consistent with official recommendations and the remainder of the documentation which shows 12 bytes for GCM.

benjreinhart committed:
Clarify initialization vector sizes in subtlecrypto docs

sideshowbarker commented:
@twiss if you could please review

benjreinhart commented:
If helpful, the [NIST publication on GCM](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf) (which is linked on this MDN docs page) says:

> For IVs, it is recommended that implementations restrict support to the length of 96 bits, to promote interoperability, efficiency, and simplicity of design.

twiss commented:
Hey :wave: Thanks! I agree the original text is wrong, but the updated text is also not quite correct. While NIST recommends 12 bytes, it doesn't require it (nor does the Web Crypto spec). Their recommendations for generating them also aren't simply that it should be "preferably cryptographically random", btw; a counter is also acceptable and preferable for IVs smaller than 12 bytes. But you could say that it should be "preferably 12 bytes of cryptographically random data", that would probably be close enough to being true :)

Also, I think it would be better to split up the section into two, one for AES-GCM and one for AES-CBC, to make it easier to state the requirements/recommendations for each. (There are also some parameters of AES-GCM missing, namely `additionalData` and `tagLength`. After splitting up the sections, those could then be added more easily later.)

Finally, the [page on `decrypt`](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) has the same mistake, you may want to change that page too? (Also, on that page it may be helpful to state that the `iv` (and `additionalData` and `tagLength`, once we add that) must be the same one that was used for encryption, and not a newly generated random one, of course.)

benjreinhart committed:
More accurate IV descr and consistency among encrypt/decrypt

benjreinhart commented:
Hey @twiss, excellent suggestions, thank you!

I took a stab at addressing your suggestions. Happy to rework it if needed.

Note that there is an [AesGcmParams](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams) page on MDN that corresponds to the spec. I pulled some language from this page to be consistent. However, in an effort to not duplicate all the content there, I linked out to that page from the encrypt/decrypt docs.

This ^ makes me think we should maybe have pages for each of the param objects and then just link out to them on the encrypt/decrypt pages?

twiss commented:
Hey @benjreinhart, thanks! This already seems like a large improvement to me.

> Note that there is an [AesGcmParams](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams) page on MDN that corresponds to the spec. I pulled some language from this page to be consistent. However, in an effort to not duplicate all the content there, I linked out to that page from the encrypt/decrypt docs.
> 
> This ^ makes me think we should have pages for each of the param objects (I see only a page for GCM params) and then just link out to those on the encrypt/decrypt pages? Otherwise, it might be a tough balancing act between documenting the necessary nuance, being concise given these params are not the page's main topic, and not wanting to duplicate too much content between the encrypt/decrypt pages as well as individual param pages.

But yeah, I tend to agree. I followed the links on the [AesGcmParams](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams) page and noticed that actually the pages on [wrapKey](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey) and [unwrapKey](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/unwrapKey) have the same mistake - so the same content is duplicated 4 times. So making pages for each of the params types and linking to them from those 4 pages seems better, indeed, if you're up for it :) The same is also done for the other operations, such as [sign](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/sign) and [verify](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/verify), so that'll make it more consistent as well.

wbamberg commented:
Thanks for your work on this PR! Actually there used to be standalone pages for the different algorithm parameters, but they were removed (incompletely apparently) in https://github.com/mdn/content/pull/21434 . This is part of a general project on MDN to avoid dictionary objects that are only used in one or two places, since they just add unnecessary indirection in a lot of cases and it seems clearer to document them in the place they are used.

But it's a balance - if a dictionary is used in a lot of places then it's better to keep it. And here where it's used in four places and is quite complicated, perhaps we made the wrong call here and should undo this.

wbamberg commented:
I just filed https://github.com/mdn/content/pull/23170 to restore the separate algorithm pages. Perhaps the best plan would be to merge that PR, and then you can repurpose this one to correct and remaining errors in these pages?

I'm sorry to put you to extra work (but it is probably less work than creating all these pages from scratch!).

benjreinhart commented:
@wbamberg sounds good to me, thanks!

hamishwillee commented:
Howdy @benjreinhart - that PR just merged, so if you want to re-integrate your changes, there is now a clean slate for the purpose.
